### DISCUSSION TITLE
In case you get this error when using useHydrateAtoms - TypeError: Cannot read properties of undefined (reading 'apply')

### DISCUSSION THREAD

**Discussion by amirulabu** [pmndrs/jotai#2237] [General] [answered]
I want to jot down this quickly just in case anyone get this error. This error happen due to you are trying to initialize an atom that is derived from other atoms. 

eg. 

This is your store
```ts
const isShowAtom = atom(true)
const amountAtom = atom(1000)

const isShowAmountAtom = atom((get)=> get(amountAtom)>100 && get(isShowAtom))
```

You cannot use `useHydrateAtoms` to directly set `isShowAmount`. you need to set `isShowAtom` and `amountAtom`

**Comment by dai-shi** [comment, marked as answer]
Yes, it's how it's designed since Jotai v2.0.0. Alternatively, you can make `isShowAmountAtom` writable.

**Reply by moelashmawy** [reply]
in case I have an atomWithReducer with complex state,

```
const initState = { count: 1,  isShow: false,  text: 'text' };

const someAtomWithReducer = atomWithReducer(initState, reducer);

const countAtom = atom((get) => get(someAtomWithReducer).count);
const isShowAtom = atom((get) => get(someAtomWithReducer).isShow);
const textAtom = atom((get) => get(someAtomWithReducer).text);
```

and in my tests
```
const HydrateAtoms = ({ initialValues, children }: any) => {
  useHydrateAtoms(initialValues);
  return children;
};

const TestProvider = ({ initialValues, children }: any) => (
  <Provider>
    <HydrateAtoms initialValues={initialValues}>{children}</HydrateAtoms>
  </Provider>
);

const mockedData = { count: 555,  isShow: true,  text: 'mocked text' };

const MockedApp = () => (
  <TestProvider
    initialValues={[
      [someAtomWithReducer, mockedData],
  >
    <Componenet Uses countAtom,  isShowAtom, textAtom/>
  </TestProvider>
);
```
injecting mockedData doesn't work and atoms return the `initState` above

it used to work fine before migrating to jotai V2, but now it doesn't work, and as I understand from the [migration guide](https://jotai.org/docs/guides/migrating-to-v2-api#some-other-changes) that useHydrateAtoms can only accept writable atoms.
isn't atomWithReducer a writable atom? and how can I achieve the tests in my case above?

**Reply by dai-shi** [reply]
Good point. This is an edge case that is no longer supported in v2.

I guess you need to separate atomWithReducer into two atoms, if you want to hydrate.

```js
const initState = { count: 1,  isShow: false,  text: 'text' };
const reducer = ...;

const baseAtom = atom(initState); // used for testing
const someAtomWithReducer = atom(
  (get) => get(baseAtom),
  (get, set, action) => set(baseAtom, reducer(get(baseAtom), action)),
);
```
